File: haskell/p08.hs
-- P08 Eliminate consecutive duplicates of list elements.

module P08 where

import Data.List

-- Built-in functions composition
f1 :: Eq a => [a] -> [a]
f1 = map head . group

-- Recursion with dropWhile
f2 :: Eq a => [a] -> [a]
f2 [] = []
f2 (x:xs) = x : (f2 $ dropWhile (== x) xs)

-- Recursion with pattern matching (look-ahead)
f3 :: Eq a => [a] -> [a]
f3 (a:(bs@(b:_)))
    | a == b = f3 bs
    | otherwise = a : f3 bs
f3 xs = xs

-- Tail recursion (look-behind)
f4 :: Eq a => [a] -> [a]
f4 = f4' []
    where f4' acc [] = reverse acc
          f4' [] (x:xs) = f4' [x] xs
          f4' ys@(y:_) (x:xs) = if y == x then f4' ys xs else f4' (x : ys) xs

-- List folding
f5 :: Eq a => [a] -> [a]
f5 [] = []
f5 xs = foldr (\a b -> if a == (head b) then b else a : b) [last xs] xs

-- Using "zip"
f6 :: Eq a => [a] -> [a]
f6 [] = []
f6 xs@(x:_) = x : map snd (filter (uncurry (/=)) (zip xs (tail xs)))

File: scala/P08.scala
// P08 Eliminate consecutive duplicates of list elements.

// Recursion with "dropWhile"
def f1[T]: List[T] => List[T] = {
  case Nil => Nil
  case h :: t => h :: f1(t.dropWhile(h ==))
}

// Tail recursion with "dropWhile"
@annotation.tailrec
def f2[T](list: List[T], acc: List[T] = Nil): List[T] = list match {
  case Nil => acc.reverse
  case h :: t => f2(t.dropWhile(h ==), h :: acc)
}

// Recursion with pattern matching
def f3[T]: List[T] => List[T] = {
  case a :: (bs@(b :: _)) => if (a == b) f3(bs) else a :: f3(bs)
  case xs => xs
}

// Folding
def f4[T]: List[T] => List[T] = {
  case Nil => Nil
  case l => l.foldRight(List(l.last))((b, a) => if (a.head == b) a else b :: a)
}

// Using "zip"
def f5[T]: List[T] => List[T] = {
  case Nil => Nil
  case xs@(x :: _) => x :: xs.zip(xs.tail).filter(p => p._1 != p._2).map(_._2)
}